airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py

```diff
@@ -247,6 +251,31 @@ def ti_run(
         )
 
 
+def _get_upstream_map_indexes(
+    request: Request, dag_id: str, task_id: str, ti_map_index
+) -> dict[str, int | list[int] | None] | None:
+    dag = request.app.state.dag_bag.get_dag(dag_id)
+    if not dag:
+        return None
+
+    task = dag.get_task(task_id)
+    upstream_map_indexes = {
+        upstream_task.task_id: (
+            # regular tasks
+            None
+            if upstream_task.task_group is None
+            # tasks in the same mapped task group
+            else ti_map_index
+            if task.task_group == upstream_task.task_group
+            # tasks not in the same mapped task group
+            # the upstream mapped task group should combine the xcom as a list and return it
+            else list(range(upstream_task.task_group.get_parse_time_mapped_ti_count()))
+        )
+        for upstream_task in task.upstream_list
+    }
+    return upstream_map_indexes
```

Comment:
This would be a lot nicer to read if implemented like this instead:

```python
def _get_upstream_map_indexes(task, ti_map_index) -> Iterator[tuple[str, int | list[int] | None]]:
    for upstream_task in task.upstream_list:
        if upstream_task.task_group is None:
            yield upstream_task.task_id, None
        elif ...


upstream_map_indexes = (
    dict(_get_upstream_map_index(dag.get_task(ti.task_id), ti.map_index)
    if dag := request.app.state.dag_bag.get_dag(ti.dag_id)
    else None
)
```